feature: support imports from azure devops

andrei-ivanovici opened the pull request:
## Hey, I just made a Pull Request!

Adds support for Azure  DevOps as a catalog import source. This enables Backstage to create PRs to Azure DevOps as it does with GitHub . It will use the logged in user as the author of the PR
There are no breaking changes

**NOTE:**  
  This feature uses the Azure REST APIs instead  of the `azure-devops-node-api` pacakge . This is because the  package does not work in the browser. 
[This](https://github.com/microsoft/azure-devops-node-api/issues/575)  issue was created but as of now  it doesn't appear to be a priority. 

![image](https://github.com/backstage/backstage/assets/8040066/6d7b0a46-47c3-44fa-a689-06ff286f1594)


![image](https://github.com/backstage/backstage/assets/8040066/1ebfa9f3-df52-4d09-8e4e-e829829757aa)

#### :heavy_check_mark: Checklist



<!--- Please include the following in your Pull Request when applicable: -->

- [x ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ x] Tests for new functionality and regression tests for bug fixes
- [x ] Screenshots attached (for UI changes)
- [x ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

andrei-ivanovici commented on plugins/catalog-import/src/api/AzureRepoApiClient.test.ts:
```diff
@@ -0,0 +1,362 @@
+  const server = setupServer();
+  setupRequestMockHandlers(server);
+  const sut = new RepoApiClient({
+    token: 'token',
```
This is a random string value used in unit tests

andrei-ivanovici commented on plugins/catalog-import/src/api/AzureRepoApiClient.test.ts:
```diff
@@ -0,0 +1,362 @@
+      fileContent: 'content',
+      branchName: 'backstage-integration',
+      description: 'Test Description',
+      token: 'token',
```
This is a random string value used in unit tests

andrei-ivanovici commented on plugins/catalog-import/src/components/ImportInfoCard/ImportInfoCard.tsx:
```diff
@@ -77,10 +79,18 @@ export const ImportInfoCard = (props: ImportInfoCardProps) => {
             Link to a repository{' '}
-            <Chip label="GitHub only" variant="outlined" size="small" />
+            <Chip
+              label="GitHub and AzureDevops"
```
Those are the only two implementations available now.  I am happy to change this to whatever makes sense

andrei-ivanovici commented:
> This PR has been automatically marked as stale because it has not had recent activity from the author. It will be closed if no further activity occurs. If the PR was closed and you want it re-opened, let us know and we'll re-open the PR so that you can continue the contribution!

This requires a review.  There are no further changes that need adding unless the reviewers request it

awanlin commented:
Will review, just got missed as it came in just when my monthly long travels ended and was catching up. 👍

awanlin commented on .changeset/rude-buckets-invite.md:
```diff
@@ -0,0 +1,7 @@
+
+Integrated Azure Devops as a catalog import source. This enables Backstage to create Pull Requests to Azure Devops repositories as it does with GitHub repositories
+
+NO BREAKING CHANGES
```
If there are no breaking changes then this doesn't need to be added

awanlin commented on .changeset/rude-buckets-invite.md:
```diff
@@ -0,0 +1,7 @@
+'@backstage/plugin-catalog-import': patch
+---
+
+Integrated Azure Devops as a catalog import source. This enables Backstage to create Pull Requests to Azure Devops repositories as it does with GitHub repositories
```
```suggestion
Integrated Azure DevOps as a catalog import source. This enables Backstage to create Pull Requests to Azure DevOps repositories as it does with GitHub repositories
```

awanlin commented on plugins/catalog-import/src/api/AzureRepoApiClient.ts:
```diff
@@ -0,0 +1,269 @@
+  };
+}
+
+const apiVersions = {
```
We currently support `6.0` of the Azure DevOps API - https://github.com/backstage/backstage/blob/d26e5a5d45ee34d3633df8f878477ccd22d947ae/packages/integration/src/azure/AzureUrl.ts#L138 - we should target that here as well.

awanlin replied:
Sorry, I should have been more clear. Your changes should be updated to support version `6.0` of the Azure DevOps API. There are Backstage Adopters using this version and will expect this feature to work for them as every other area of Backstage that supports Azure DevOps works with their version of the API.

Are you sure the version `6.0` for these end points doesn't work? It's been a long time but I have high confidence that this works as I've done it in PowerShell before just using the REST endpoints. 🤔

andrei-ivanovici replied:
I have tested the endpoints with Versions prior to 7 and not all  appeared to work, I'm afraid :( . So  I narrowed them down the common denominator. I can look a bit more, maybe I will find alternatives  with V6
Will come back with updates and maybe some references

andrei-ivanovici replied:
Ok. I was wrong. It appears to be working perfectly fine with V6. I will make the changes and update the PR

awanlin reviewed (changes requested):
Thanks for this contribution @andrei-ivanovici, took a quick look and overall this seems great to me. Left a couple of comments. I plan to try this out ended to end late next week.

awanlin commented on plugins/catalog-import/src/components/ImportInfoCard/ImportInfoCard.tsx:
```diff
@@ -40,7 +41,8 @@ export interface ImportInfoCardProps {
     exampleLocationUrl = 'https://github.com/backstage/backstage/blob/master/catalog-info.yaml',
-    exampleRepositoryUrl = 'https://github.com/backstage/backstage',
+    exampleGitRepositoryUrl = 'https://github.com/backstage/backstage',
+    exampleAzureRepositoryUrl = 'https://dev.azure.com/spotify/backstage/_git/backstage',
```
```suggestion
    exampleAzureRepositoryUrl = 'https://dev.azure.com/org-name/project-name/_git/repo-name,
```

awanlin commented on plugins/catalog-import/src/components/ImportInfoCard/ImportInfoCard.tsx:
```diff
@@ -77,10 +79,18 @@ export const ImportInfoCard = (props: ImportInfoCardProps) => {
           </Typography>
           <Typography variant="subtitle2" color="textSecondary" paragraph>
-            Example: <code>{exampleRepositoryUrl}</code>
+            Azure Example: <code>{exampleAzureRepositoryUrl}</code>
```
```suggestion
            Azure DevOps Example: <code>{exampleAzureRepositoryUrl}</code>
```

awanlin replied:
Should we also include an example for Azure DevOps Server?

`https://azure-devops.mycompany.com/team-collection-name/team-project-name/_git/repo-name`

freben replied:
Another way to go would be to revert the changes here instead - this doesn't really scale; it's just examples anyway. If the original props were all we had, integrators could choose to put other examples in there that are perfect for their specific org. Not sure we should be extending the props here tbh

awanlin replied:
Sure, I'm good with that! 🚀

andrei-ivanovici replied:
Thank you for coming back to me on this. Are we happy with the suggested change here ? Or do we want to revert the component's props to their original values  ? Happy to go with whatever you guys  feel is more appropriate

Sorry if I'm asking to many questions :)

awanlin replied:
Thanks for reverting 👍

awanlin commented on plugins/catalog-import/src/api/AzureDevops.ts:
```diff
@@ -0,0 +1,81 @@
+  repoUrl: string,
+  integration: AzureIntegration,
+): AzureRepoParts {
+  const { organization, owner, name } = parseGitUrl(repoUrl);
```
We've had issues in the past using `git-url-parse` it doesn't play nicely with Azure DevOps Server as it can be hosted on a sub-path, this was the common pattern for TFS ( the previous name for Azure DevOps). This is what we have done to get around this: https://github.com/backstage/community-plugins/blob/a25418c4f52b759f78a1e4256b28f891e673f334/workspaces/azure-devops/plugins/azure-devops-backend/src/utils/azure-devops-utils.ts#L332

awanlin replied:
There's a similar function here: https://github.com/backstage/backstage/blob/4eca697bf3c5a996795324444f605b918f4491be/packages/integration/src/azure/AzureUrl.ts#L25 

I'm not sure how usable that would be here though 🤔

andrei-ivanovici replied:
Interesting ... My expectation was that the `git-url-parser` would handle that on its own. I'm glad we caught this.
At first glance your first suggestion would work nicely but  from what I could tell it is part of a plugin.  What is the recommended approach here ? copy the function to the catalog-import  plugin ? or move that function to a common  package and reference it in both places ?  Adding a dependency to the plugin and using it doesn't feel right.

 I will pick this up when I return from vacation towards the end of the week.

Thank you !

andrei-ivanovici replied:
I  have applied the previously discussed changes 
  - Reverted the examples to default
  - Changed the  URL parser to use the one from azure-devops-backend. I did this by copying the function across.  Adding a dependency didn't feel right.
  
  **NOTE**: I tried to use `AzureUrl` class from `intergrations` but found out that it does not play nice with TFS servers.  The parsing logic will result to the owner being wrongly set to `tfs/organisation_name` rather than `organisation_name`
Might be worth considering to unify these in the integrations package

Looking forward to your feedback

awanlin replied:
This looks great 👍

awanlin reviewed (approved):
This looks great to me now, feel like we should ship it an iterate. 🚀 

Though there is props to set the examples in the `ImportInfoCard` there's no way for an Adopter to actually use them, that would be a great follow up PR to expose them.

awanlin commented:
@freben ready for your review when you have time, please. 😄

awanlin commented:
Hi @camilaibs @freben, this is good to go, can you give it a review please?

Phiph commented:
Hey I'd love to see this in an upcoming version to help speed-up the onboarding of my organisation!

andrei-ivanovici commented on plugins/catalog-import/src/api/util.ts:
```diff
@@ -0,0 +1,40 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+export function parseRepoUrl(sourceUrl: string) {
```
Thanks for looking over this. As per @awanlin's recommendations, git-url-parser is not appropriate here as it does not cater for the tfs self hosted scenario.  I can move this  closer to where it is used.

andrei-ivanovici commented on plugins/catalog-import/src/api/AzureRepoApiClient.ts:
```diff
@@ -0,0 +1,266 @@
+    const endpoint = this.createEndpoint(path, version, queryParams);
+    const result = await fetch(endpoint, {
+      headers: {
+        Authorization: `Bearer ${this._options.token}`,
```
The client is  meant to  be transient. The usage pattern would  be: create a new instance, use it, then dispose of  it.  I can change this to get the token as a parameter for all operations if that feels  more natural :)

andrei-ivanovici commented on plugins/catalog-import/src/api/CatalogImportClient.ts:
```diff
@@ -144,7 +146,7 @@ export class CatalogImportClient implements CatalogImportApi {
     return {
       type: 'repository',
-      integrationType: 'github',
+      integrationType: foundIntegration!.type,
```
`isSupported` captures the undefined case  but the compiler does not know that. I  will include that check to make it happy

awanlin commented:
Hi @andrei-ivanovici, just did a quick look and some of the comments aren't resolved?

@freben, looks like we might need your thoughts on some of them, please.

andrei-ivanovici commented:
> Hi @andrei-ivanovici, just did a quick look and some of the comments aren't resolved?
> 
> @freben, looks like we might need your thoughts on some of them, please.

Do you mind sharing which of the  points haven't been addressed? 
As far as I can tell all   points have changes associated with them:
 - Moved the parser code closer  to where it's used and remove the util file
 - Moved the token and is now a parameter in the client operations
 - Removed the exclamation mark by updating the code so that the compiler catches that use-case
 - Updated the error message to me account for multiple providers
 
If there is anything else that I'm missing feel free to share.

awanlin commented on plugins/catalog-import/src/api/CatalogImportClient.ts:
```diff
@@ -89,15 +89,17 @@ export class CatalogImportClient implements CatalogImportApi {
+          `The ${foundIntegration.title} integration only supports full URLs to ${catalogFilename} files. Did you try to pass in the URL of a directory instead?`,
         );
       }
       throw new Error(
```
@andrei-ivanovici, this was the comment I thought wasn't addressed but looking at it seems like it has, thanks 👍

freben reviewed (approved):
Let's go!
